File: Python/PROGRAMMERS/CODINGTEST_PRACTICE/network.py
def solution(n, computers):
    def find(x):
        if x == par[x]:
            return x

        par[x] = find(par[x])
        return par[x]

    def _union(x, y):
        x = find(x)
        y = find(y)

        par[x] = y
        
    # computers 는 인접행렬
    
    par = list(range(n + 1))
    
    # 인접리스트로 연결시키기
    for i in range(n):
        for j in range(i, n):
            if computers[i][j]:
                if find(i) == find(j):
                    continue
                _union(i, j)
                
    # 경로압축으로 부모 일치시키기
    for i in range(n):
        find(i)
    
    # 중복 제거
    par = set(list(par[:-1]))
                
    return len(par)

File: C++/PROGRAMMERS/network.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int par[201];

int find(int x) {
    if(par[x] == x)
        return x;
    
    return par[x] = find(par[x]);
}

void _union(int a,int b) {
    a = find(a);
    b = find(b);

    par[a] = b;
}

int solution(int n, vector<vector<int>> computers) {

    for(int i = 0; i < n; i++)
        par[i] = i;
    
    for(int i = 0; i < n; i++)
    {
        for(int j = 0; j < n; j++)
        {
            if(computers[i][j]==1)
                _union(i,j);
        }
    }
    for(int i = 0; i < n; i++)
        par[i] = find(i);

    sort(par, par + n);
    
    int temp = par[0], cnt = 1;
    for(int i = 1; i < n; i++) {
        if(par[i] != temp) {
            temp = par[i];
            cnt++;
        }
    }
    
    return cnt;
}